electrum/lnworker.py

```diff
@@ -2062,10 +2049,7 @@ async def create_routes_for_payment(
                                 partial(
                                     self.create_route_for_single_htlc,
                                     amount_msat=part_amount_msat,
-                                    invoice_pubkey=paysession.invoice_pubkey,
-                                    min_final_cltv_delta=paysession.min_final_cltv_delta,
-                                    r_tags=paysession.r_tags,
-                                    invoice_features=paysession.invoice_features,
+                                    paysession=paysession,
```

Comment:
Note that the fields of `PaySession` are mutable.  We are passing them to a function that is run asynchronously (`await run_in_thread`). This happens to be fine in that case, because the fields we are passing (`invoice_pubkey, invoice, r_tags, min_final_cltv_delta`) happen to not be mutated during the lifetime of the session. However, this seems to be lucky. What would happen if we had been passing `paysession._amount_inflight,` or some other variable field? Then the value at the time it is evaluated in the coroutine might differ from the value passed as parameter.

Reply:
We could use [dataclasses](https://docs.python.org/3/library/dataclasses.html) for `PaySession` and force immutability using the `frozen` flag

Reply:
`NamedTuple` seems to be too limited in this case AFAICS

Reply:
We could store the relevant (immutable) fields in a `NamedTuple` within `PaySession` and pass that instead. This way there is no access to the mutable fields of `PaySession` in the routing methods

Reply:
> Note that the fields of PaySession are mutable

Good point.

> store the relevant (immutable) fields in a NamedTuple within PaySession

(btw besides NamedTuple, there is also `@dataclass(frozen=True)` in the [stdlib](https://docs.python.org/3/library/dataclasses.html). It is similar to attrs, a tiny bit less flexible, but attrs I would eventually like to remove from the dependencies)

Reply:
> (btw besides NamedTuple, there is also `@dataclass(frozen=True)` in the [stdlib](https://docs.python.org/3/library/dataclasses.html). It is similar to attrs, a tiny bit less flexible, but attrs I would eventually like to remove from the dependencies)

Yes I have looked at [dataclasses](https://docs.python.org/3/library/dataclasses.html), but this is too inflexible, as there are mutable fields in `PaySession` that should remain mutable, and I haven't seen an option to selectively un-immute (heh) a subset of fields. Subclassing also doesn't allow mutability as the restriction propagates down the hierarchy.

Reply:
I added a method to return an immutable copy and pass that instead. See https://github.com/spesmilo/electrum/pull/9623/commits/dc1e5a024432325e583077b81553e964d1940435

There's probably nicer ways to do this, e.g. externalize the `__setattr__` override

Reply:
Hmm.. I'm not really happy with the last commit. If we copy the PaySession it does not really need to be immutable, and the added code is.. not very nice.

Maybe it is nicer to add a `NamedTuple` within `PaySession`, and just pass that immutable tuple down to the routing functions.

Reply:
indeed I would prefer not to pass `paysession` to that method, but a namedtuple that can be a field of `paysession`.

The other call (`create_trampoline_route_and_onion`) needs to pass a mutable field (`failed_trampoline_routes`) in addition to immutable fields. That mutable field should be passed explicitly, in addition to the namedtuple.

Reply:
as I'm reworking the patch to pass a `NamedTuple`, it looks like there are too many exceptions to really benefit from this encapsulation :(

passing the `NamedTuple` and extra fields of `PaySession` along is not really an improvement of the old situation IMO.